### PR TITLE
 CBG-355: External config for cache enhancements

### PR DESCRIPTION
**Pull request by JRascagneres**
- [x] Merge https://github.com/couchbaselabs/sync-gateway-accel/pull/234
- [x] Bump manifest accordingly

### CODE REVIEW

File: rest/config.go [hunk 1]
```diff
@@ -256,13 +259,34 @@ type EventConfig struct {
...
 
 type CacheConfig struct {
-	CachePendingSeqMaxWait *uint32 `json:"max_wait_pending,omitempty"` // Max wait for pending sequence before skipping
-	CachePendingSeqMaxNum  *int    `json:"max_num_pending,omitempty"`  // Max number of pending sequences before skipping
-	CacheSkippedSeqMaxWait *uint32 `json:"max_wait_skipped,omitempty"` // Max wait for skipped sequence before abandoning
-	EnableStarChannel      *bool   `json:"enable_star_channel"`        // Enable star channel
-	ChannelCacheMaxLength  *int    `json:"channel_cache_max_length"`   // Maximum number of entries maintained in cache per channel
-	ChannelCacheMinLength  *int    `json:"channel_cache_min_length"`   // Minimum number of entries maintained in cache per channel
-	ChannelCacheAge        *int    `json:"channel_cache_expiry"`       // Time (seconds) to keep entries in cache beyond the minimum retained
+	RevCacheConfig     *RevCacheConfig     `json:"rev_cache"`     // Revision Cache Config Settings
```
Comment (bbrks) [inline rest/config.go]: We should probably retain the `omitempty` flag on all of the struct tags.

File: rest/config.go [hunk 2]
```diff
@@ -414,6 +438,25 @@ func (dbConfig DbConfig) validate() error {
 		}
 	}
 
+	if dbConfig.CacheConfig != nil {
+
+		if dbConfig.CacheConfig.ChannelCacheConfig != nil {
+			if dbConfig.CacheConfig.ChannelCacheConfig.MaxNumber != nil {
+				if *dbConfig.CacheConfig.ChannelCacheConfig.MaxNumber < 1 {
+					return fmt.Errorf("minimum value for max_number is 1")
```
Comment (bbrks) [inline rest/config.go]: Make the error message channel cache config specific, so the user knows exactly where this option is.

File: rest/config.go [hunk 3]
```diff
@@ -414,6 +438,25 @@ func (dbConfig DbConfig) validate() error {
...
+				}
+			}
+		}
+
+		if dbConfig.CacheConfig.RevCacheConfig != nil {
+			if dbConfig.CacheConfig.RevCacheConfig.ShardCount != nil {
+				if *dbConfig.CacheConfig.RevCacheConfig.ShardCount < 1 {
+					return fmt.Errorf("minimum value for shard_count is 1")
```
Comment (bbrks) [inline rest/config.go]: Make the error message rev cache config specific, so the user knows exactly where this option is.

**Review by bbrks — CHANGES_REQUESTED** [review]
Just a couple of small changes

File: manifest/default.xml [hunk 4]
```diff
@@ -25,7 +25,7 @@
 
   
   <!-- Sync Gateway Accel-->
-  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="d379b87427b107a01f3cf48121642c8e52aa227a"/>
+  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="6115c2aa1b625df2a9a3c9fe723305a30667b727"/>
```
Comment (bbrks) [inline manifest/default.xml]: Merged revision: `d1fa2f687b65e8ca103ce7f2764cb06c7c6c196f`

```suggestion
  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="d1fa2f687b65e8ca103ce7f2764cb06c7c6c196f"/>
```

File: manifest/default.xml [hunk 5]
```diff
@@ -25,7 +25,7 @@
 
   
   <!-- Sync Gateway Accel-->
-  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="d379b87427b107a01f3cf48121642c8e52aa227a"/>
+  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="d1fa2f687b65e8ca103ce7f2764cb06c7c6c196f"/>
```
Comment (bbrks) [inline manifest/default.xml]:
```suggestion
  <project groups="notdefault,sg-accel" name="sync-gateway-accel" path="godeps/src/github.com/couchbaselabs/sync-gateway-accel" remote="couchbaselabs_private" revision="b8ac98d63dd3b34c5c2c062ec293fd8df193fca2"/>
```

File: rest/config.go [hunk 6]
```diff
@@ -256,13 +259,34 @@ type EventConfig struct {
...
+
+type ChannelCacheConfig struct {
+	MaxNumber              *int    `json:"max_number,omitempty"`
+	MaxWaitPending         *uint32 `json:"max_wait_pending,omitempty"`
+	MaxNumPending          *int    `json:"max_num_pending,omitempty"`
+	MaxWaitSkipped         *uint32 `json:"max_wait_skipped,omitempty"`
+	EnableStarChannelCache *bool   `json:"enable_star_channel_cache,omitempty"`
+	MaxLength              *int    `json:"max_length,omitempty"`
+	MinLength              *int    `json:"min_length,omitempty"`
+	Expiry                 *int    `json:"expiry,omitempty"`
```
Comment (bbrks) [inline rest/config.go]: Can we retain the comments describing what these fields do?

File: rest/config.go [hunk 7]
```diff
@@ -256,13 +259,34 @@ type EventConfig struct {
...
+	// CacheSkippedSeqMaxWait *uint32 `json:"max_wait_skipped,omitempty"` // Max wait for skipped sequence before abandoning
+	// EnableStarChannel      *bool   `json:"enable_star_channel"`        // Enable star channel
+	// ChannelCacheMaxLength  *int    `json:"channel_cache_max_length"`   // Maximum number of entries maintained in cache per channel
+	// ChannelCacheMinLength  *int    `json:"channel_cache_min_length"`   // Minimum number of entries maintained in cache per channel
+	// ChannelCacheAge        *int    `json:"channel_cache_expiry"`       // Time (seconds) to keep entries in cache beyond the minimum retained
+}
+
+type RevCacheConfig struct {
+	Size       *uint32 `json:"size,omitempty"`
+	ShardCount *uint16 `json:"shard_count,omitempty"`
```
Comment (bbrks) [inline rest/config.go]: Similarly, add a comment to each describing what the field is used for.

File: rest/config.go (hunk 6 above)
Comment (bbrks) [inline rest/config.go]: Would it make sense to rename to include seconds as the units now?

**Review by bbrks — CHANGES_REQUESTED** [review]

File: rest/config.go [hunk 8]
```diff
@@ -256,13 +259,34 @@ type EventConfig struct {
...
+}
+
+type ChannelCacheConfig struct {
+	MaxNumber              *int    `json:"max_number,omitempty"`
+	MaxWaitPending         *uint32 `json:"max_wait_pending,omitempty"`
+	MaxNumPending          *int    `json:"max_num_pending,omitempty"`
+	MaxWaitSkipped         *uint32 `json:"max_wait_skipped,omitempty"`
+	EnableStarChannelCache *bool   `json:"enable_star_channel_cache,omitempty"`
```
Comment (adamcfraser) [inline rest/config.go]: As discussed, leave as enable_star_channel (remove 'cache')

**Review by adamcfraser — CHANGES_REQUESTED** [review]

**Review by bbrks — APPROVED** [review]
Changes look OK to me

**Review by adamcfraser — APPROVED** [review]

**Pull request merged**